ssl/quic/quic_wire.c

```diff
@@ -0,0 +1,776 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/macros.h>
+#include <openssl/objects.h>
+#include "quic_local.h"
+#include "internal/quic_vlint.h"
+#include "internal/quic_wire.h"
+
+/*
+ * QUIC Wire Format Encoding
+ * =========================
+ */
+
+int ossl_quic_wire_encode_frame_padding(WPACKET *pkt, size_t num_bytes)
+{
+    /*
+     * PADDING is frame type zero, which as a variable-length integer is
+     * represented as a single zero byte. As an optimisation, just use memset.
+     */
+    return WPACKET_memset(pkt, 0, num_bytes);
+}
+
+static int encode_frame_hdr(WPACKET *pkt, uint64_t frame_type)
+{
+    return WPACKET_quic_write_vlint(pkt, frame_type);
+}
+
+int ossl_quic_wire_encode_frame_ping(WPACKET *pkt)
+{
+    return encode_frame_hdr(pkt, OSSL_QUIC_FRAME_TYPE_PING);
+}
+
+int ossl_quic_wire_encode_frame_ack(WPACKET *pkt,
+                                    uint32_t ack_delay_exponent,
+                                    const OSSL_ACKM_ACK *ack)
```

Comment:
I agree, but I think it's the most practical choice. The structure would be exactly the same if we redefined it, as in both cases we want a logical representation of the ACK frame contents. There's also a dependency question here; should the wire format PR depend on the ACKM PR or vice versa? If people want the struct defined in this PR and we want to merge this before merging ACKM I can do that. Actually this would have the benefit of making this PR not depend on ACKM. If people are more likely to review this before ACKM, that may be a good idea.

Reply:
I think it is a bad "smell" indicating that we don't quite understand how the layering is working. What we seem to have here is QUIC wire format encoding/decoding component and an ACK Manager component. The dependency is from the wire format encoder/decoder to the ACK Manager, i.e. the encoder/decoder knows about the ACK Manager, not the other way around.

The design in this PR is the other way around for all the other frame types, i.e. the encoder/decoder component doesn't know anything about any other components, it defines its own structures and expects other components to use them. I think this asymmetry is purely an artefact of the way we have started building the code. We wrote the ACK Manager first, and haven't built other components yet.

One possibility is to move the "ownership" of the ACK structure into the encoder/decoder. That would be most consistent with what you've got for all the other frame types - but feels a bit wrong. In my mind encoding/decoding is a very low-level concept - so having it own the data structures for all the rest of the code feels off.

Another way to design this is to not have the encoding/decoding as a separate component at all. Instead it could be split up into separate pieces, e.g. so that the ACK Manager knows how to do encoding/decoding for ACK frames. Other components know how to do encoding/decoding for their respective frame types. In terms of this PR that might mean moving the ACK encoding/decoding out of quic_wire.c and into quic_ackm.c. We don't have components for the other frame types yet - but we could move the relevant functions as we create them (i.e. defer to later PRs).

Reply:
I think it would be reasonable to move the ACK structure into this PR and rebase the ACKM PR on top of it. In fact, it's probably a good idea. Wire format encoding is a utility and it makes sense for ACKM to depend on it rather than vice versa.

It would also be possible to have a PR just with struct definitions, and treat the logical representation of the different frame types as something to be committed in itself, depended on by both this and the ACKM PR. We've hit similar issues recently with depending on types for QUIC packet numbers, time representations (which was originally part of Pauli's event queue work but which I also rely on), etc.

Though, where code lives after it's committed is orthogonal to how work is split up into PRs. In practice I think the structures may as well be committed together as part of this PR. That doesn't mean that those structures need to be considered "part" of the wire format encoding functionality. C isn't really modular anyway, all we can really do is decide what header file something lives in.

I think it would be a very bad idea to put ACKM in charge of ACK serialization. These were separated for a reason. The user of the ACKM isn't obliged to send every range described logical ACK provided by the ACKM even, if the number of ranges is high. They can choose what to pack into a packet. The ACKM has high inherent complexity already and moving serialization into it would enhance its complexity and reduce its testability, and the ability for the TXP to take the discretionary decisions it is entitled to when deciding how much space in a packet to allocate to ACK frames.

Reply:
> I think it would be reasonable to move the ACK structure into this PR and rebase the ACKM PR on top of it. In fact, it's probably a good idea. Wire format encoding is a utility and it makes sense for ACKM to depend on it rather than vice versa.

I can live with that.